### PR TITLE
Update contributor docs, add maintainer docs

### PR DESCRIPTION
**Pull request by mcmire**
I started with adding a section on releases and then I realized that there's a lot of knowledge in my head that I haven't really shared, so I did as much brain-dumping as I could. I know this is a lot of information, take your time in reviewing it. But I at least pointed out where the release instructions are so you can have that info now. There's also some info around generating documentation too that you might find helpful.

---

* Move section on documentation from README to CONTRIBUTING
* Add a setup script that contributors and maintainers can use to set up
  a dev environment

[skip ci]

### CODE REVIEW

File: MAINTAINING.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+[semver]: https://semver.org/spec/v2.0.0.html
+
+* We bump the "major" part of the version if we're introducing
+  backward-incompatible changes (e.g. changing the API or core behavior,
+  removing parts of the API, or dropping support for a version of Ruby).
+* We bump the "minor" part if we're adding a new feature (e.g. adding a new
+  matcher or adding a new qualifier to a matcher).
+* We bump the "patch" part if we're merely including bugfixes.
+
+### Releasing a new version
```
Comment (mcmire) [inline MAINTAINING.md]: Here's the part about releasing a new version :)

File: MAINTAINING.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+## Addendum: Labels
+
+In order to corral the issue and PR backlog, we've found
+[labels] to be useful for cataloguing and tracking progress purposes. Over time
+we've added quite a collection of labels. Here's a quick list:
+
+[labels]: https://github.com/thoughtbot/shoulda-matchers/labels
+
+### Labels for issues
```
Comment (mcmire) [inline MAINTAINING.md]: I added a section for labels as a sort of brain dump. These labels aren't set in stone, of course, and I realize it's entirely possible we have too many. I'd be curious to know what your thoughts on them are.

File: MAINTAINING.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+* We bump the "minor" part if we're adding a new feature (e.g. adding a new
+  matcher or adding a new qualifier to a matcher).
+* We bump the "patch" part if we're merely including bugfixes.
+
+In addition to major, minor, and patch levels, you can also append a
+suffix to the version for pre-release versions. We usually use this to issue
+release candidates prior to an actual release. A version number in this case
+might look like `4.0.0.rc1`.
+
+### Releasing a new version
```
Comment (mcmire) [inline MAINTAINING.md]: Here's the part about releasing a new version :)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,92 +1,172 @@
...
+which are old and some of which are new. The best approach for writing tests is
+probably to copy an existing test in the same file as where you want to add a
+new test.
 
-   In any case, to run the entire test suite again, say:
+## Code style
 
-      bundle exec rake
+We follow a derivative of the [unofficial Ruby style guide] created by the
+Rubocop developers. You can view our Rubocop configuration [here], but here are
```
Comment (aried3r) [inline CONTRIBUTING.md]: Nitpick: It's "RuboCop" (see https://github.com/rubocop-hq/rubocop).

**Pull request merged**